src/librustc/lint/builtin.rs

```diff
@@ -2083,6 +2097,20 @@ impl LintPass for PrivateNoMangleFns {
                     cx.span_lint(PRIVATE_NO_MANGLE_FNS, it.span, msg.as_slice());
                 }
             },
+            ast::ItemStatic(..) => {
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") &&
+                       !cx.exported_items.contains(&it.id) {
+                    let msg = format!("static {} is marked #[no_mangle], but not exported",
+                                      it.ident);
+                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, msg.as_slice());
+                }
+            },
+            ast::ItemConst(..) => {
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") {
+                    let msg = "const items should never be #[no_mangle]";
```

Comment:
Sure. How concerned should I be about overly long messages? (The current
ones all seem pretty terse).

I think this should probably have a hint that "maybe you wanted pub static"
too

On Tuesday, February 3, 2015, Keegan McAllister <notifications@github.com>
wrote:

> In src/librustc/lint/builtin.rs
> <https://github.com/rust-lang/rust/pull/21881#discussion_r24023185>:
>
> > @@ -2083,6 +2097,20 @@ impl LintPass for PrivateNoMangleFns {
> >                      cx.span_lint(PRIVATE_NO_MANGLE_FNS, it.span, msg.as_slice());
> >                  }
> >              },
> > +            ast::ItemStatic(..) => {
> > +                if attr::contains_name(it.attrs.as_slice(), "no_mangle") &&
> > +                       !cx.exported_items.contains(&it.id) {
> > +                    let msg = format!("static {} is marked #[no_mangle], but not exported",
> > +                                      it.ident);
> > +                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, msg.as_slice());
> > +                }
> > +            },
> > +            ast::ItemConst(..) => {
> > +                if attr::contains_name(it.attrs.as_slice(), "no_mangle") {
> > +                    let msg = "const items should never be #[no_mangle]";
>
> Can you add a few words about why?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/21881/files#r24023185>.
>